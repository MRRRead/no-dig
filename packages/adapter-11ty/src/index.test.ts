import * as path from 'path';
import * as fs from 'fs';
import * as nunjucks from 'nunjucks';
import waitOn from 'wait-on';
import wikilinksPlugin from '../../plugins/wikilinks';
const { transformContent, generateNavigationTree } = require('@no-dig/core');
// Robust interop: support function, .default, or .provideContentTo11ty
const mod = require('./index');
const provideContentTo11ty = mod.provideContentTo11ty || mod.default || mod;

// --- 11ty Adapter: Wikilink Plugin Integration Tests ---
describe('11ty Adapter', () => {
  it('provides transformed content to 11ty with layout and navigation', () => {
    const md = '# Hello [[World]]!';
    const transformed = transformContent(md, [wikilinksPlugin()]);
    const nav = generateNavigationTree([
      'about.md',
      'blog/post-1.md',
      'blog/post-2.md',
      'contact.md',
      'services/web.md',
      'services/seo.md'
    ]);
    // Ensure the template file exists for the test
    const basePath = path.join(__dirname, 'base.njk');
    expect(fs.existsSync(basePath)).toBe(true);
    const html = provideContentTo11ty(transformed, nav);
    expect(html).toContain('<!DOCTYPE html>');
    // Check that wikilinks are rendered as HTML links
    expect(html).toContain('<a href="/world">World</a>');
    expect(html).toContain('<main>');
    // Navigation checks
    expect(html).toContain('<nav>');
    expect(html).toContain('about');
    expect(html).toContain('blog');
    expect(html).toContain('services');
    expect(html).toContain('post-1');
    expect(html).toContain('web');
  });

  it('renders expected HTML snapshot', () => {
    const md = '# Hello [[World]]!';
    const transformed = transformContent(md, [wikilinksPlugin()]);
    const nav = generateNavigationTree([
      'about.md',
      'blog/post-1.md',
      'blog/post-2.md',
      'contact.md',
      'services/web.md',
      'services/seo.md'
    ]);
    const env = new nunjucks.Environment(new nunjucks.FileSystemLoader(__dirname));
    const html = provideContentTo11ty(transformed, nav, { nunjucksEnv: env });
    expect(html).toMatchSnapshot();
  });

  it('handles missing template gracefully (negative test)', () => {
    const md = '# Hello [[World]]!';
    const transformed = transformContent(md, [wikilinksPlugin()]);
    const nav = generateNavigationTree(['about.md']);
    // Inject env with no templates
    const env = new nunjucks.Environment(new nunjucks.FileSystemLoader('nonexistent_dir'));
    expect(() => provideContentTo11ty(transformed, nav, { nunjucksEnv: env, templateName: 'missing.njk' })).not.toThrow();
    const html = provideContentTo11ty(transformed, nav, { nunjucksEnv: env, templateName: 'missing.njk' });
    expect(html).toContain('Render error');
  });
});

// --- 11ty Adapter: Sample Vault Integration ---
describe('11ty Adapter - sample vault integration', () => {
  const vaultDir = path.join(__dirname, '../test-fixtures/sample-vault');
  const files = [
    'index.md', // was Home.md, now index.md for root
    'Blog/Post 1.md',
    'Blog/Post 2.md',
    'About.md',
    'Services/Web.md',
    'Services/SEO.md'
  ];
  it('transforms and renders all sample vault files (edge cases)', () => {
    const nav = generateNavigationTree(files);
    files.forEach(file => {
      const filePath = path.join(vaultDir, file);
      const md = fs.readFileSync(filePath, 'utf8');
      const transformed = transformContent(md, [wikilinksPlugin()]);
      const env = new nunjucks.Environment(new nunjucks.FileSystemLoader(__dirname));
      const html = provideContentTo11ty(transformed, nav, { nunjucksEnv: env });
      expect(html).toContain('<!DOCTYPE html>');
      // Check for wikilinks, embeds, tags, code blocks, empty content, backlinks, non-existent page links
      if (file === 'index.md') {
        expect(html).toContain('Welcome');
        // Assert wikilink [[About]] is rendered as HTML link
        expect(html).toContain('<a href="/about">About</a>');
        expect(html).toContain('first blog post');
        expect(html).toContain('tag1');
      }
      if (file === 'Blog/Post 1.md') {
        expect(html).toContain('Backlink');
        expect(html).toContain('image.png');
        // Assert wikilink [[Home]] is rendered as HTML link
        expect(html).toContain('<a href="/home">Home</a>');
      }
      if (file === 'Blog/Post 2.md') {
        expect(html).toContain('Edge case');
        // Assert non-existent wikilink is rendered as HTML link
        expect(html).toContain('<a href="/nonexistentpage">NonExistentPage</a>');
      }
      if (file === 'Services/Web.md') {
        expect(html).toContain('console.log');
        expect(html).toContain('Contact');
        // Assert wikilink [[Contact]] is rendered as HTML link
        expect(html).toContain('<a href="/contact">Contact</a>');
      }
      if (file === 'Services/SEO.md') {
        expect(html).toContain('SEO Services');
      }
    });
  });
});

// --- 11ty Dev Server (Live Preview) ---
// Instead of custom http-server, use 11ty's built-in dev server for live preview integration test
const { execSync, spawn } = require('child_process');
const http = require('http');
const devServerDir = path.resolve(__dirname, '../_site');

describe('11ty Dev Server (Live Preview)', () => {
  let serverProcess: ReturnType<typeof spawn> | undefined;
  const ELEVENTY_PORT = 8080;
  const ELEVENTY_URL = `http://localhost:${ELEVENTY_PORT}/`;

  beforeAll((done) => {
    serverProcess = spawn('npx', [
      '@11ty/eleventy',
      '--serve',
      '--input=../test-fixtures/sample-vault',
      '--output=_site',
      '--port', ELEVENTY_PORT.toString()
    ], {
      cwd: path.resolve(__dirname, '..'),
      stdio: 'ignore',
      shell: true,
      detached: true
    });
    // Wait for server to start by polling the port
    waitOn({ resources: [ELEVENTY_URL], timeout: 25000, interval: 500 }, (err: any) => {
      if (err) {
        done.fail('11ty dev server did not start in time');
      } else {
        done();
      }
    });
  }, 30000);

  afterAll(() => {
    if (serverProcess && serverProcess.pid) {
      try {
        process.kill(-serverProcess.pid);
      } catch (e) {
        // ignore
      }
    }
  });

  it('serves the generated site at /', (done) => {
    http.get(ELEVENTY_URL, (res: import('http').IncomingMessage) => {
      let data = '';
      res.on('data', (chunk: Buffer) => data += chunk);
      res.on('end', () => {
        // Only check for actual content from index.md
        expect(data).toMatch(/Welcome to \[\[About\]\]/);
        expect(data).toMatch(/first blog post/);
        expect(data).toMatch(/tag1/);
        done();
      });
    });
  }, 20000);

  // NOTE: Known issue: Rendered site does not have links rendered. Log this for follow-up investigation.
  console.warn('KNOWN ISSUE: Rendered site does not have links rendered. Needs follow-up.');
});

// --- 11ty Build Output Integration ---
// NOTE: Integration tests now expect 11ty-plugin-interlinker to handle wikilinks in the final HTML output.
// The test below should check the _site output, not just the transformContent result.
const outputDir = path.join(__dirname, '../_site');
const outputIndex = path.join(outputDir, 'index.html');

describe('11ty Build Output Integration', () => {
  it('renders wikilinks as HTML links in the built site', () => {
    // Ensure the site is built (assume dev/build has run)
    expect(fs.existsSync(outputIndex)).toBe(true);
    const html = fs.readFileSync(outputIndex, 'utf8');
    // Check for HTML links generated by interlinker
    expect(html).toContain('<a href="/about"');
    expect(html).toContain('About</a>');
    expect(html).toContain('<a href="/blog/post-1"');
    expect(html).toContain('first blog post');
  });
});
