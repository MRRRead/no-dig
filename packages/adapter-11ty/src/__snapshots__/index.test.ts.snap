// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`11ty Adapter renders expected HTML snapshot 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>NO-DIG Site</title>
</head>
<body>
  <nav>
    <ul>
      
        <li>
          
            <a href="/about">about</a>
          
          
        </li>
      
        <li>
          
            blog
          
          
            <ul>
              
                <li>
                  <a href="/blog/post-1">post-1</a>
                </li>
              
                <li>
                  <a href="/blog/post-2">post-2</a>
                </li>
              
            </ul>
          
        </li>
      
        <li>
          
            <a href="/contact">contact</a>
          
          
        </li>
      
        <li>
          
            services
          
          
            <ul>
              
                <li>
                  <a href="/services/web">web</a>
                </li>
              
                <li>
                  <a href="/services/seo">seo</a>
                </li>
              
            </ul>
          
        </li>
      
    </ul>
  </nav>
  <main>
    # Hello <a href="/world">World</a>!
  </main>
</body>
</html>
"
`;
